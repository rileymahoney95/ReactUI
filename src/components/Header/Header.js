import React from 'react';
import './Header.scss'

const Header = props => {

    function logout(){
        window.location = props.ssoInfo.logoutUrl;
    }

    return (
        <div className="header">
            <div className="header-logo">
                <i className="icon_homedepot"></i>
            </div>
            <div className="header-info">
                <label className="product-info">Supply Chain Network Flow Path</label>
                <label className="user-info">{props.name} </label>
            </div>
                <div className="header-actions">
                <button id="logoutBtn" className="button primary sm" onClick={logout}>Logout</button>
            </div>
        </div>
    );
}

export default Header;