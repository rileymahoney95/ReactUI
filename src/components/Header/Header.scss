.header-actions {
    margin-left: auto;
    padding-right: 20px;
    padding-top: 10px;
}