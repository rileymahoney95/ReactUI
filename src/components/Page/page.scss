.page {
  position: relative;
  top: 120px;
  text-align: center;
}
#responseBanner {
  text-align: center;
}
#form {
  position: relative;
  right: 65px;
}
#dropzone {
  display: block;
  margin: auto auto;
  width: 260px;
  height: 200px;
  border: 5px dashed;
}
#dzText {
  position: relative;
  top: 80px;
}
#filesToUpload {
  position: relative;
  right: 74px;
}
#check {
  color: green;
}
#download {
  display: flex;
  justify-content: center;
}

.select-container::before {
  z-index: 100;
  top: 1.3em;
  right: 0.7em;
}

.select-container select {
  padding: 0 10px;
  margin-right: 10px;
  font-size: 14px;
  min-height: 30px;
}

.select-container {
  margin: 10px;
}

.select-container:first-of-type {
  margin: 10px;
}

.select-container:last-of-type {
  margin: 10px;
}