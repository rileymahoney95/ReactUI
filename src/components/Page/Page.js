import React, { Component } from 'react';
import Dropzone from 'react-dropzone';
import axios from 'axios'

import './page.scss'

export default class Page extends Component {
    constructor(props) {
        super(props);
        
        this.state = {
            files: [],
            bannerType: "",
            bannerMessage: "",
            uploadDisabled: true,
            templateDropdown: "df_upload"
        }

        this.templates = {
            df_upload: "SupplyChainNetworkFlowPath_Template.csv",
            flowpath_upload: "Flowpath_Template.csv"
        }
        this.closeBanner = this.closeBanner.bind(this);
        this.showResponse = this.showResponse.bind(this);
        this.handleDrop = this.handleDrop.bind(this);
        this.handleUploadClick = this.handleUploadClick.bind(this);
        this.onChange = this.onChange.bind(this);
    }

    handleDrop (files) {
        this.setState({
            bannerType: "",
            bannerMessage: "",
            files: files,
            uploadDisabled: false
        });
    }

    tokenHasExpired() {
        var now = Date.now() / 1000;
        return now > this.props.expireTime;
    }

    tokenSignatureIsValid() {
        return this.props.verifyToken(this.props.id_token, this.props.publicKey) ? true : false;
    }

    handleUploadClick () {
        if(this.tokenHasExpired() || !this.tokenSignatureIsValid()) {
            return this.props.updateSession(false);
        }
        var formData  = new FormData();
        var statusCode = 0;

        formData.append("directFulfillmentSKUNetworkUpload", this.state.files[0]);

        axios.post(`${this.props.flowPathUrl}?userID=${this.props.ldap}&email=${this.props.email}`, formData)
            .then((response) =>  {
                statusCode = response.status;
                this.showResponse(response.data, statusCode);
            })
            .catch((error) => {
                console.error(`${error}`);
                this.showResponse(error.response.data, statusCode);
            })
     }

    showResponse(response, status) {
        let bannerType;
        let bannerMessage;
        if (status === 200) {
            if (response.invalidRecords === 0) {
                bannerType = 'banner success';
                bannerMessage = response.message;
            } else {
                bannerType = 'banner warning';
                bannerMessage = `${response.message}. Invalid Records: ${response.invalidRecords}`;
            }
        } else {
            bannerType = 'banner error';
            bannerMessage = response.message;
        }
        this.setState({
            files: [],
            uploadDisabled: true,
            bannerType: bannerType,
            bannerMessage: bannerMessage
        });
    }

    closeBanner() {
        this.setState({
            bannerType: '',
            bannerMessage: ''
        });
    }

    onChange(e) {
        this.setState({templateDropdown: e.target.value});
    }
    

    render() {
        return (
            <div className="page">
                <div id="responseBanner" className={this.state.bannerType}>
                    <span id="responseMessage" className="message">{this.state.bannerMessage}</span>
                    <span onClick={this.closeBanner} className="close"></span>
                </div>

                <Dropzone className="dz-box" id="dropzone" onDrop={this.handleDrop} accept=".csv">
                    <div id="dzText" className="dz-message" data-dz-message>
                        <span>Select or Drop File</span>
                    </div>
                </Dropzone>

                <p id="filesToUpload"><b>File to Upload</b></p> 

                <p>
                    <i id="check" className="icon_check-circle-filled" hidden={!this.state.files[0]}></i>
                    {this.state.files[0] ? this.state.files[0].name : "No file selected"}
                </p>
                <div id="download">
                    <div className="select-container">
                        <select onChange={this.onChange} value={this.state.templateDropdown} className="dropdown-list sm">
                            <option value="df_upload">Direct Fulfillment Upload</option>
                            <option value="flowpath_upload">Flowpath Upload</option>
                        </select>
                    </div>
                    
                    <a href={`../${this.templates[this.state.templateDropdown]}`} download target='_blank'>
                        <button title="Download Button" className="button primary sm">Download Template</button>
                    </a>

                    <span onClick={this.handleUploadClick}>
                        <button id="btnUpload" className="button primary sm" disabled={this.state.uploadDisabled}>Upload</button>
                    </span>
                </div>
            </div>
        );
    }
}