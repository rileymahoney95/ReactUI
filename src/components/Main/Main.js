import React, { Component } from 'react'
import { Redirect } from 'react-router-dom'
import axios from 'axios'
import jwt from 'jsonwebtoken'

import Header from '../Header/Header'
import Page from '../Page/Page'

export default class Main extends Component {
    constructor(props) {
        super(props);

        this.state = {
            ldap: '',
            name: '',
            expireTime: null,
            sessionValid: this.isSessionValid(),
            publicKey: '',
            id_token: '',
            email: ''
        };
        this.updateSession = this.updateSession.bind(this);
    }

    componentDidMount() {
        if (!this.isSessionValid()) return; // used to terminate execution when page is redirected back to home route
        axios.get(this.props.config.ssoInfo.tokenKeysUrl).then((res) => {
            const key = res.data.keys[0].value;
            const id_token = window.location.hash.split('&')[2].split('=')[1];
            this.props.history.replace('/callback')
            const decoded_token = this.verifyAndDecodeToken(id_token, key);
            if (decoded_token) {
                const email = decoded_token.email;
                const name = email.slice(0, email.indexOf('@')).split('_').join(' ');
                const ldap = decoded_token.user_name;
                const expireTime = decoded_token.exp;
                this.setState({
                    ldap: ldap,
                    name: name,
                    expireTime: expireTime,
                    publicKey: key,
                    id_token: id_token,
                    email: email
                });
            }
        }).catch((err) => {
            console.error(`Error getting public key for token: ${err}`);
        });
    }

    verifyAndDecodeToken(id_token, key) {
        let decoded_token = null;
        try {
            decoded_token = jwt.verify(id_token, key, { algorithms: ['RS256'] });
        } catch (error) {
            console.error(`Error verifying and/or decoding token: ${error}`);
        }
        return decoded_token;
    }

    isSessionValid() {
        return window.location.hash.length > 0 && 
          window.location.hash.includes('access_token') &&
          window.location.hash.includes('token_type=bearer')
    }

    updateSession(value) {
        this.setState({
            sessionValid: value
        });
    }

    render() {
        if (!this.state.sessionValid) {
            return <Redirect to='/' />
        }
        return (
            <div>
                <Header name={this.state.name} ssoInfo={this.props.config.ssoInfo}/>
                <Page ldap={this.state.ldap} email={this.state.email} expireTime={this.state.expireTime} verifyToken={this.verifyAndDecodeToken} publicKey={this.state.publicKey} 
                      id_token={this.state.id_token} updateSession={this.updateSession} flowPathUrl={this.props.config.flowPathUrl}/>
            </div>
        );
    }
};