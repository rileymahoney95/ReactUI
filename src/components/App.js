import React from 'react';
import { Router, Route, Switch } from 'react-router-dom'
import 'babel-polyfill';
import axios from 'axios';

import SSO from './SSO/SSO';
import Main from './Main/Main';
import { getSSOInfo } from '../util/ssoHelper';

class App extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            config: {}
        }
    }

    componentDidMount() {
        axios.get('/config').then(res => {
            const info = getSSOInfo(res.data);
            this.setState({
                config: {
                    ssoInfo: info,
                    flowPathUrl: res.data.FLOW_PATH_URL
                }
            });
        }).catch(err => {
            console.error(`Error getting config: ${err}`)
        });
    }

    render() {
        if (Object.keys(this.state.config).length === 0) {
            return (
                <div className="progress-linear indeterminate"></div>
            )
        }
        else{
            return (
                <Router history={this.props.history}>
                    <Switch>
                        <Route exact path='/' render={() => <SSO config={this.state.config} />} />
                        <Route path='/callback' render={(props) => <Main {...props} config={this.state.config} />} />
                    </Switch>
                </Router>
            );
        }
    }
}

export default App;

