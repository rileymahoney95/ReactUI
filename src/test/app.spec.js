import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import './setupTests.js';
import axios from 'axios';

import App from '../components/App';
import Main from '../components/Main/Main';
import Page from '../components/Page/Page';
import Header from '../components/Header/Header';
import { createBrowserHistory } from 'history';

jest.mock("axios");

describe("<App /> Component", () => {
    it('renders App without crashing', () => {
        const mockConfigData = {   
        data: { 
            SSO_URL: "test", 
            SSO_CLIENT_ID: "1234", 
            FLOW_PATH_URL: "https://example.com" 
            }
        };
        axios.get = () => new Promise(resolve => resolve(mockConfigData));
        const history = createBrowserHistory();
        const output = shallow(
            <App history={history}/>
        );
        expect(shallowToJson(output)).toMatchSnapshot();
    });
});

describe("<Main /> Component", () => {
    it('renders Main without crashing', () => {
        const output = shallow(
            <Main />
        );
        expect(shallowToJson(output)).toMatchSnapshot();
    });

    
});

describe("<Header /> Component", () => {
    it('renders Header without crashing', () => {
        const output = shallow(
            <Header />
        );
        expect(shallowToJson(output)).toMatchSnapshot();
    });
});

describe("<Page /> Component", () => {

    it('renders Page without crashing', () => {
        const output = shallow(
            <Page />
        );
        expect(shallowToJson(output)).toMatchSnapshot();
    });

    it('enables upload button on file drop', () => {
        const page = shallow(
            <Page />
        );
        expect(page.state().uploadDisabled).toEqual(true);
        page.instance().handleDrop(page.state().files);
        expect(page.state().uploadDisabled).toEqual(false);
    });

    it('disables upload button on upload button click', () => {
        const page = shallow(
            <Page />
        );
        page.setState({
            uploadDisabled: false
        });
        page.instance().showResponse({ invalidRecords: 0, message: "Your upload has been submitted.", validRecords: 3 }, 200);
        expect(page.state().uploadDisabled).toEqual(true);
    });

    it('shows success banner for file with all valid records', () => {
        const page = shallow(
            <Page />
        );
        page.instance().showResponse({ invalidRecords: 0, message: "Your upload has been submitted.", validRecords: 3 }, 200);
        expect(page.state().bannerType).toBe("banner success");
    });

    it('shows warning banner for file with some valid records', () => {
        const page = shallow(
            <Page />
        );
        page.instance().showResponse({ invalidRecords: 2, message: "Your upload has been submitted.", validRecords: 1 }, 200);
        expect(page.state().bannerType).toBe("banner warning");
    });

    it('shows error banner for file with no valid records', () => {
        const page = shallow(
            <Page />
        );
        page.instance().showResponse({ message: "Error: Cannot parse uploaded file." }, 422);
        expect(page.state().bannerType).toBe("banner error");
    });

});