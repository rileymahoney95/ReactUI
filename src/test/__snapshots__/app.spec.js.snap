// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> Component renders App without crashing 1`] = `
<div
  className="progress-linear indeterminate"
/>
`;

exports[`<Header /> Component renders Header without crashing 1`] = `
<div
  className="header"
>
  <div
    className="header-logo"
  >
    <i
      className="icon_homedepot"
    />
  </div>
  <div
    className="header-info"
  >
    <label
      className="product-info"
    >
      Supply Chain Network Flow Path
    </label>
    <label
      className="user-info"
    >
       
    </label>
  </div>
  <div
    className="header-actions"
  >
    <button
      className="button primary sm"
      id="logoutBtn"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
</div>
`;

exports[`<Main /> Component renders Main without crashing 1`] = `
<Redirect
  push={false}
  to="/"
/>
`;

exports[`<Page /> Component renders Page without crashing 1`] = `
<div
  className="page"
>
  <div
    className=""
    id="responseBanner"
  >
    <span
      className="message"
      id="responseMessage"
    />
    <span
      className="close"
      onClick={[Function]}
    />
  </div>
  <t
    accept=".csv"
    className="dz-box"
    disableClick={false}
    disablePreview={false}
    disabled={false}
    id="dropzone"
    inputProps={Object {}}
    maxSize={Infinity}
    minSize={0}
    multiple={true}
    onDrop={[Function]}
    preventDropOnDocument={true}
  >
    <div
      className="dz-message"
      data-dz-message={true}
      id="dzText"
    >
      <span>
        Select or Drop File
      </span>
    </div>
  </t>
  <p
    id="filesToUpload"
  >
    <b>
      File to Upload
    </b>
  </p>
  <p>
    <i
      className="icon_check-circle-filled"
      hidden={true}
      id="check"
    />
    No file selected
  </p>
  <div
    id="download"
  >
    <a
      download={true}
      href="../SupplyChainNetworkFlowPath_Template.csv"
      target="_blank"
    >
      <button
        className="button primary sm"
        title="Download Button"
      >
        Download Template
      </button>
    </a>
    <span
      onClick={[Function]}
    >
      <button
        className="button primary sm"
        disabled={true}
        id="btnUpload"
      >
        Upload
      </button>
    </span>
  </div>
</div>
`;
